pypsa/optimization/optimize.py

```diff
@@ -104,6 +104,14 @@ def define_objective(n, sns):
         object_const = m.add_variables(constant, constant, name="objective_constant")
         objective.append(-1 * object_const)
 
+    # spill cost
+    for c, attr in lookup.index[lookup.index.get_level_values(1) == "spill_cost"]:
+        cost = get_as_dense(n, c, "spill_cost", sns).loc[:, lambda ds: (ds != 0).any()]
+        if cost.empty:
+            continue
+        spill = m[f"{c}-spill"].sel({"snapshot": sns, c: cost.columns})
+        objective.append((spill * cost).sum())
```

Comment:
This does not consider snapshot weightings or period weightings.

It's not consistent with the unit of the `spill` variable, which is MW, whereas the cost unit is `currency/MWh.

I think the easiest way to implement it would be  to wrap the marginal cost block in a loop:

```py
for cost_type in {"marginal_cost", "spill_cost"}:
    for c, attr in lookup.query(cost_type).index:
        cost = (
            get_as_dense(n, c, cost_type, sns)
            .loc[:, lambda ds: (ds != 0).any()]
            .mul(weighting, axis=0)
        )
        if cost.empty:
            continue
        operation = m[f"{c}-{attr}"].sel({"snapshot": sns, c: cost.columns})
        objective.append((operation * cost).sum())
```

Reply:
Hi Fneum, thanks for the comments I will go ahead with those changes now.
On another note, I am kind of confused why the code coverage check is complaining because I added a unit test which does run the new code. Could you help me understand why that is happening?

Reply:
Also in terms of the units I am a bit confused. It appears that the marginal cost (currency/MWh) is also multiplied with power (MW). Is there a hidden assumption that snapshots last one hour?

Reply:
@RobbieKiwi 
Just FYI: Code coverage check was fixed in #1014. If you rebase and trigger the CI again, the report should be fixed.

Reply:
@lkstrp 
I am still having issues with the code coverage. I wrote some new lines which are covered but I think I deleted a few more (covered) lines than I wrote, thereby reducing total coverage by 0.02% and failing the coverage check